proposals/0000-split-wunused-imports.md

```diff
@@ -0,0 +1,230 @@
+
+0. Whenever a user has `-Wunused-imports` on, and they don't have `-Wno-meta-unused-imports` on, emit a warning that highlights the changed behaviour of `-Wunused-imports`.
+1. It is on by default.
+2. It **is not made into an error by `-Werror`**. Optionally there could be `-Wmeta-error` or something along those lines for users who want to get errors for potential future meta warnings.
+
+This `-Wmeta-unused-improts` warning is only present for a few release cycles, and is removed afterwards, or alternatively, turned off by default.
+
+## Examples
+
+###
+
+```haskell
+import Foo
+import Foo (x)
+
+bla = x
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+
+Proposed:
+* with `-Wunused-imports` - nothing
+* with `-Wduplicate-imports` - warn that the `Foo` import is duplicate
+
+###
+
+```haskell
+import Foo (x)
+import Bar (x)
+
+bla = x
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Bar` import is unused
+
+Proposed:
+* with `-Wunused-imports` - nothing
+* with `-Wduplicate-imports` - warn that the `Bar` import is duplicate
+
+###
+
+```haskell
+import Foo
+import Bar
+
+bla = x
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Bar` import is unused
+
+Proposed:
+* with `-Wunused-imports` - nothing
+* with `-Wduplicate-imports` - warn that the `Bar` import is duplicate
+
+###
+
+```haskell
+import Foo
+
+bla = ()
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+
+Proposed:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+* with `-Wduplicate-imports` - nothing
+
+###
+
+```haskell
+import Foo (x)
+
+bla = ()
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+
+Proposed:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+* with `-Wduplicate-imports` - nothing
+
+## Effect and Interactions
+
+Unsure what to fill in here, it seems that the [Proposed Change Specification](#proposed-change-specification) covers the effects of this change.
+
+## Costs and Drawbacks
+
+The main cost is changing the behaviour of a warning without notice, even if we explicitly warn users that it has changed.
+
+Is this acceptable?
+
+We're in the camp of "yes, it is", **if** we include the meta warning, since anyone who is maintaining code and cares about keeping things warning free should be able to
+easily notice the meta warning and turn on `-Wduplicate-imports` (or turn off the meta warning).
+
+There's also the cost of introducing the new concept of a "meta warning". These could arguably warrant their own standalone proposal, where they are more precisely specified.
+
+## Alternatives
+
+### Don't include the meta warning and let users deal with the silently changed behaviour.
+This seems potentially frustrating. Who knows if someone has some niche use case that they really like for the "duplicate imports" part of `-Wunused-imports`?
```

Comment:
Unless we have actual evidence that this change would be harmful, I'd be very reluctant to introduce such a new concept. Feels like vorauseilender Gehorsam.

Reply:
If I'm correctly following what "vorauseilender Gehorsam" means, I also speak partly for myself here - while this change in particular is not something that I would mind, I am not a fan of things changing silently under my feet.

I agree that it's a bit heavyweight to introduce an entirely new "type" of output (albeit very similar to existing ones) from GHC, I also feel that it might have been helpful in the past or be helpful in the future in any similar "transition periods".

Alternatively, I remember some discussion about extending/changing warning categories, and perhaps even excluding some warning categories from being errors by default. It might make sense to turn this into a category - "meta warnings" - that are not errors with `-Werror`. That way, this can fit into pre-existing concepts. I can't remember where this discussion was being held though.

Reply:
Thanks for writing up this proposal! It's a good idea to tackle this issue.

> Alternatively, I remember some discussion about extending/changing warning categories, and perhaps even excluding some warning categories from being errors by default. It might make sense to turn this into a category - "meta warnings" - that are not errors with `-Werror`. That way, this can fit into pre-existing concepts. I can't remember where this discussion was being held though.

There was some discussion in #541 about being able to classify the default severity of custom warnings from `WARNING` pragmas, perhaps that's what you are thinking of? I don't recall any discussion of errors that would not become warnings with `-Werror`.

I'm also pretty sceptical of the value of `-Wmeta-unused-imports`. It will be a very noisy way to herald the change, seems difficult to disable (since `-Wno-meta-unused-imports` will give an "unrecognised warning flag" warning on earlier compilers), and introduces complexity to the definition of `-Werror`. Warnings can and do change between GHC releases, so a mention in the release notes seems like sufficient notice to me.

Reply:
> since -Wno-meta-unused-imports will give an "unrecognised warning flag" warning on earlier compilers

I completely forgot this part. I admit that it does indeed greatly reduce my enthusiasm for the meta warning.